In [1]:
import numpy as np
import pandas as pd
import gc
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split
import pickle


#Graph


from sklearn.preprocessing import normalize
from sklearn.preprocessing import scale, robust_scale, minmax_scale, maxabs_scale
import lightgbm as lgb

from sklearn.preprocessing import StandardScaler 
from sklearn import linear_model, decomposition, datasets
from sklearn.pipeline import Pipeline
from sklearn.model_selection import GridSearchCV
from sklearn.decomposition import PCA
from sklearn.neighbors import KNeighborsClassifier # KNN
from sklearn.ensemble import RandomForestClassifier # RandomForest
from sklearn.model_selection import cross_val_score #cross validaiton
from sklearn import cross_validation
from sklearn.preprocessing import scale, robust_scale, minmax_scale, maxabs_scale # Scaling
from sklearn.model_selection import KFold
import xgboost as xgb
from sklearn.model_selection import KFold
from sklearn.model_selection import GridSearchCV 
from sklearn.metrics import f1_score
from sklearn.feature_selection import RFE
from sklearn.svm import SVC
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.feature_selection import SelectFromModel
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import roc_auc_score

/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/sklearn/cross_validation.py:44: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.
  "This module will be removed in 0.20.", DeprecationWarning)


In [2]:
pd.set_option('display.max_columns', 150)

# Trainset Import

In [3]:
X = pickle.load(open( "pickle/loan2", "rb" ) )

In [4]:

X.drop(['index', 'id'], axis=1, inplace =True)

## To solve the unbalanced data, dataset will be separated by label(y)

### (1) X_1 : 'y' value is 1 (defaul)

In [5]:
X_1 =X.loc[X.loan_status==1]

In [6]:
X_1 = X_1.reset_index()
X_1.drop('index', axis = 1, inplace = True)
X_1.head()

loan_amnt  funded_amnt  funded_amnt_inv  int_rate  installment  sub_grade  \
0    15250.0      15250.0          15250.0     13.11       514.65   0.828769   
1    25000.0      25000.0          25000.0     19.05       917.04   0.649399   
2    35000.0      35000.0          35000.0     24.08      1008.51   0.560944   
3    14000.0      14000.0          13975.0     11.99       464.94   0.848689   
4     4000.0       4000.0           4000.0     12.99       134.76   0.778796   

   emp_title  emp_length  annual_inc  desc    dti  delinq_2yrs  \
0          1           5     55000.0     1  15.86          0.0   
1          1           3    125000.0     1  22.38          0.0   
2          1          10    114000.0     1  28.15          0.0   
3          1           8     39000.0     1  19.75          0.0   
4          1           1     18900.0     1   9.67          0.0   

   inq_last_6mths  mths_since_last_delinq  mths_since_last_record  open_acc  \
0             0.0               70.000000                74.31392       8.0   
1             1.0               28.000000                74.31392      14.0   
2             2.0               31.000000                74.31392      10.0   
3             0.0               34.887939               101.00000      11.0   
4             0.0               34.887939                74.31392      17.0   

   pub_rec  revol_bal  revol_util  total_acc  collections_12_mths_ex_med  \
0      0.0    17098.0        70.9       22.0                         0.0   
1      0.0    28852.0        86.1       24.0                         0.0   
2      0.0    40583.0        90.2       47.0                         0.0   
3      1.0    13986.0        58.0       21.0                         0.0   
4      0.0     4631.0        49.3       19.0                         0.0   

   mths_since_last_major_derog  acc_now_delinq  tot_coll_amt  tot_cur_bal  \
0                         43.0             0.0           0.0     281777.0   
1                         43.0             0.0         108.0      64649.0   
2                         43.0             0.0           0.0     284638.0   
3                         43.0             0.0           0.0     118782.0   
4                         43.0             0.0           0.0      55474.0   

   open_acc_6m  open_il_6m  open_il_12m  open_il_24m  mths_since_rcnt_il  \
0          0.0         0.0          0.0          0.0                 0.0   
1          0.0         0.0          0.0          0.0                 0.0   
2          0.0         0.0          0.0          0.0                 0.0   
3          0.0         0.0          0.0          0.0                 0.0   
4          0.0         0.0          0.0          0.0                 0.0   

   total_bal_il  il_util  open_rv_12m  open_rv_24m  max_bal_bc  all_util  \
0           0.0      0.0          0.0          0.0         0.0       0.0   
1           0.0      0.0          0.0          0.0         0.0       0.0   
2           0.0      0.0          0.0          0.0         0.0       0.0   
3           0.0      0.0          0.0          0.0         0.0       0.0   
4           0.0      0.0          0.0          0.0         0.0       0.0   

   total_rev_hi_lim  inq_fi  total_cu_tl  inq_last_12m  loan_status  \
0           24100.0     0.0          0.0           0.0            1   
1           33500.0     0.0          0.0           0.0            1   
2           45000.0     0.0          0.0           0.0            1   
3           24100.0     0.0          0.0           0.0            1   
4            9400.0     0.0          0.0           0.0            1   

   term_ 36 months  term_ 60 months  grade_A  grade_B  grade_C  grade_D  \
0                1                0        0        1        0        0   
1                1                0        0        0        0        1   
2                0                1        0        0        0        0   
3                1                0        0        1        0        0   
4             

In [7]:
print (X_1.shape)

(45936, 142)


### (2) X_0: 'y'value is 0 (paid)

In [8]:
X_0 = X.loc[X.loan_status == 0]
X_0 = X_0.reset_index()
X_0.drop('index', axis = 1, inplace=True)
X_0.tail()

loan_amnt  funded_amnt  funded_amnt_inv  int_rate  installment  \
157224     5000.0       5000.0           5000.0     12.39       167.01   
157225    12000.0      12000.0          12000.0     11.71       396.92   
157226    20000.0      20000.0          19800.0     12.12       665.44   
157227     8475.0       8475.0           8475.0     18.92       310.32   
157228    15000.0      15000.0          15000.0     17.10       535.54   

        sub_grade  emp_title  emp_length  annual_inc  desc    dti  \
157224   0.792376          1           8    110000.0     1  22.82   
157225   0.848689          1           2     77000.0     1  11.60   
157226   0.848689          1           5     80000.0     1  18.65   
157227   0.649399          1          10     22500.0     1  17.01   
157228   0.716210          1          10     38000.0     1  11.40   

        delinq_2yrs  inq_last_6mths  mths_since_last_delinq  \
157224          0.0             0.0               42.000000   
157225          0.0             0.0               34.887939   
157226          2.0             2.0                4.000000   
157227          0.0             0.0               34.887939   
157228          1.0             0.0                7.000000   

        mths_since_last_record  open_acc  pub_rec  revol_bal  revol_util  \
157224                74.31392       8.0      0.0     8083.0        89.8   
157225                74.31392       7.0      0.0     8377.0        63.4   
157226                74.31392      13.0      0.0     5949.0        42.5   
157227                83.00000      11.0      1.0    12770.0        62.3   
157228                74.31392       6.0      0.0     8276.0        58.3   

        total_acc  collections_12_mths_ex_med  mths_since_last_major_derog  \
157224       32.0                         0.0                         42.0   
157225        8.0                         0.0                         43.0   
157226       41.0                         0.0                          4.0   
157227       16.0                         0.0                         43.0   
157228       13.0                         0.0                         43.0   

        acc_now_delinq  tot_coll_amt  tot_cur_bal  open_acc_6m  open_il_6m  \
157224             0.0           0.0     219096.0          0.0         0.0   
157225             0.0           0.0      79066.0          0.0         0.0   
157226             0.0           0.0     182909.0          0.0         0.0   
157227             0.0         100.0      13624.0          0.0         0.0   
157228             0.0           0.0     194311.0          0.0         0.0   

        open_il_12m  open_il_24m  mths_since_rcnt_il  total_bal_il  il_util  \
157224          0.0          0.0                 0.0           0.0      0.0   
157225          0.0          0.0                 0.0           0.0      0.0   
157226          0.0          0.0                 0.0           0.0      0.0   
157227          0.0          0.0                 0.0           0.0      0.0   
157228          0.0          0.0                 0.0           0.0      0.0   

        open_rv_12m  open_rv_24m  max_bal_bc  all_util  total_rev_hi_lim  \
157224          0.0          0.0         0.0       0.0       9000.000000   
157225          0.0          0.0         0.0       0.0      29543.592078   
157226          0.0          0.0         0.0       0.0      14000.000000   
157227          0.0          0.0         0.0       0.0      20500.000000   
157228          0.0          0.0         0.0       0.0      14200.000000   

        inq_fi  total_cu_tl  inq_last_12m  loan_status  term_ 36 months  \
157224     0.0          0.0           0.0            0                1   
157225     0.0          0.0           0.0            0                1   
157226     0.0          0.0           0.0            0                1   
157227     0.0          0.0           0.0            0                1   
157228     0.0          0.0           0.0            0             

In [9]:
print (X_0.shape)

(157229, 142)


In [10]:
X_0.sample(n=45936)

loan_amnt  funded_amnt  funded_amnt_inv  int_rate  installment  \
145520     6000.0       6000.0           6000.0      9.91       193.35   
16851     15000.0      15000.0          15000.0     23.10       423.72   
56404     16000.0      16000.0          16000.0     12.99       539.03   
51800     28000.0      28000.0          28000.0      8.90       889.09   
40954      5100.0       5100.0           5100.0      7.90       159.59   
33895      6000.0       6000.0           6000.0     13.53       203.70   
16661      9000.0       9000.0           9000.0     11.99       298.89   
131957    20000.0      20000.0          20000.0     17.10       498.13   
39443     23825.0      23825.0          23825.0     20.20       633.88   
13481     12000.0      12000.0          12000.0      8.90       381.04   
153671    19500.0      19500.0          19500.0     11.67       644.61   
139411    20000.0      20000.0          19975.0     16.29       706.01   
95679      5000.0       5000.0           5000.0     10.15       161.69   
23709      6575.0       6575.0           6575.0     13.11       221.89   
152178    10525.0      10525.0          10525.0     18.55       383.42   
81933     11300.0      11300.0          11300.0     12.85       379.93   
1704      12000.0      12000.0          12000.0     13.98       279.10   
19392     30000.0      30000.0          30000.0     21.18       814.65   
155317    14000.0      14000.0          14000.0     14.65       482.92   
119044     5000.0       5000.0           5000.0     11.67       165.29   
12913     20000.0      20000.0          19975.0      6.03       608.72   
155528     2800.0       2800.0           2800.0     13.99        95.69   
128590    13600.0      13600.0          13600.0     17.77       490.11   
90065      4000.0       4000.0           4000.0      5.42       120.64   
17600      9600.0       9600.0           9600.0     13.99       328.06   
85125     10000.0      10000.0          10000.0     12.85       336.22   
122155    35000.0      35000.0          35000.0     15.31      1218.61   
73186      2000.0       2000.0           2000.0     15.62        69.94   
127414     9000.0       9000.0           9000.0     12.99       303.21   
1619       4000.0       4000.0           3975.0      8.39       126.07   
...           ...          ...              ...       ...          ...   
136503    21925.0      21925.0          21925.0     24.99       643.40   
91868      2400.0       2400.0           2400.0      8.90        76.21   
11957      6850.0       6850.0           6850.0     15.31       238.50   
135770    12000.0      12000.0          11925.0      8.90       381.04   
119473    18000.0      18000.0          17950.0     24.70       525.17   
48105      5000.0       5000.0           5000.0      6.99       154.37   
64941     10000.0      10000.0          10000.0      7.69       311.94   
154206     5550.0       5550.0           5550.0      7.62       172.95   
79679      7500.0       7500.0           7500.0     11.49       247.29   
154057    14400.0      14400.0          14400.0     15.22       500.74   
79836      6300.0       6300.0           6225.0     13.72       145.68   
87510     25875.0      25875.0          25875.0     23.50       736.89   
57781     24250.0      24250.0          24250.0      8.90       770.02   
137970     3000.0       3000.0           3000.0     13.11       101.25   
105826    20000.0      20000.0          20000.0     21.15       542.76   
123293     2500.0       2500.0           2500.0      5.42        75.40   
3625      15000.0      15000.0          15000.0     12.99       505.34   
35260     18000.0      18000.0          18000.0      7.62       560.91   
108291    28000.0      28000.0          28000.0     15.10       667.59   
50774     10000.0      10000.0          10000.0     16.99       248.48   
18862      7500.0       7500.0           7500.0     20.99       282.53   
154572    24000.0      14575.0          14450.0      9.99       309.61   
56436     10

### Test set Import

In [11]:
X_test = pickle.load(open( "pickle/X_test_transformed", "rb" ) )


In [12]:
y_test = X_test['loan_status']

In [13]:
X_test.drop(['index','id','loan_status'], axis=1, inplace = True)

In [15]:
X_test.shape

(67722, 141)

In [14]:
X_test.head()

loan_amnt  funded_amnt  funded_amnt_inv  int_rate  installment  sub_grade  \
0    12000.0      12000.0          12000.0     19.72       316.06   0.645764   
1     3000.0       3000.0           3000.0     14.09       102.67   0.814785   
2     8000.0       8000.0           8000.0     16.99       285.19   0.692728   
3    12800.0      12800.0          12800.0     10.99       419.00   0.862483   
4    14000.0      14000.0          14000.0     13.11       472.46   0.828769   

   emp_title  emp_length  annual_inc  desc    dti  delinq_2yrs  \
0          1          10     46400.0     1  25.37          0.0   
1          1           5     70000.0     1  13.53          0.0   
2          1           1     45000.0     1  10.21          0.0   
3          1          10     90000.0     1  24.03          0.0   
4          1          10     55000.0     1  20.68          0.0   

   inq_last_6mths  mths_since_last_delinq  mths_since_last_record  open_acc  \
0             0.0               34.887939                74.31392      18.0   
1             2.0               49.000000                74.31392      15.0   
2             0.0               34.887939                74.31392       6.0   
3             1.0               34.887939                74.31392      15.0   
4             0.0               44.000000                74.31392      11.0   

   pub_rec  revol_bal  revol_util  total_acc  collections_12_mths_ex_med  \
0      0.0    10761.0        69.4       30.0                         0.0   
1      0.0     8919.0        44.2       24.0                         0.0   
2      0.0    10331.0        47.6       10.0                         0.0   
3      0.0    21370.0        61.1       35.0                         0.0   
4      0.0    15090.0        85.3       31.0                         0.0   

   mths_since_last_major_derog  acc_now_delinq  tot_coll_amt  tot_cur_bal  \
0                         43.0             0.0           0.0     172387.0   
1                         43.0             0.0           0.0     181147.0   
2                         43.0             0.0           0.0      10331.0   
3                         43.0             0.0           0.0     202866.0   
4                         46.0             0.0           0.0      27520.0   

   open_acc_6m  open_il_6m  open_il_12m  open_il_24m  mths_since_rcnt_il  \
0          0.0         0.0          0.0          0.0                 0.0   
1          0.0         0.0          0.0          0.0                 0.0   
2          0.0         0.0          0.0          0.0                 0.0   
3          0.0         0.0          0.0          0.0                 0.0   
4          0.0         0.0          0.0          0.0                 0.0   

   total_bal_il  il_util  open_rv_12m  open_rv_24m  max_bal_bc  all_util  \
0           0.0      0.0          0.0          0.0         0.0       0.0   
1           0.0      0.0          0.0          0.0         0.0       0.0   
2           0.0      0.0          0.0          0.0         0.0       0.0   
3           0.0      0.0          0.0          0.0         0.0       0.0   
4           0.0      0.0          0.0          0.0         0.0       0.0   

   total_rev_hi_lim  inq_fi  total_cu_tl  inq_last_12m  term_ 36 months  \
0           15500.0     0.0          0.0           0.0                0   
1           20200.0     0.0          0.0           0.0                1   
2           21700.0     0.0          0.0           0.0                1   
3           35000.0     0.0          0.0           0.0                1   
4           17700.0     0.0          0.0           0.0                1   

   term_ 60 months  grade_A  grade_B  grade_C  grade_D  grade_E  grade_F  \
0                1        0        0        0        1        0        0   
1                0        0        1        0        0        0        0   
2                0        0        0        0        1        0        0   
3                0        0        1        0        0       

# 1. Scaling
* Prepare scaled dataset as well

In [82]:
scaler = StandardScaler()
scaler.fit(X)
X_scaled = scaler.transform(X)
X_scaled=pd.DataFrame(X_scaled, columns= X.columns)

X_scaled.tail()

loan_amnt  funded_amnt  funded_amnt_inv  int_rate  installment  \
203160  -0.200927    -0.196540        -0.181824 -0.494339    -0.093247   
203161   0.778221     0.785033         0.771206 -0.401455     0.999881   
203162  -0.632365    -0.629045        -0.612520  1.139069    -0.445790   
203163   0.166253     0.171550         0.184726  0.726752     0.471066   
203164   0.288647     0.294247         0.306909  0.273657     0.552037   

        sub_grade  emp_title  emp_length  annual_inc  desc       dti  \
203160   0.683873   0.242877   -1.014276    0.075504   0.0 -0.649112   
203161   0.683873   0.242877   -0.184402    0.125288   0.0  0.248443   
203162  -1.138398   0.242877    1.198721   -0.828912   0.0  0.039650   
203163  -0.527485   0.242877    1.198721   -0.571693   0.0 -0.674574   
203164   0.044784   0.242877    1.198721    0.092099   0.0  0.660936   

        delinq_2yrs  inq_last_6mths  mths_since_last_delinq  \
203160    -0.340316       -0.769016                0.000000   
203161     2.297017        0.969918               -2.108283   
203162    -0.340316       -0.769016                0.000000   
203163     0.978351       -0.769016               -1.903515   
203164    -0.340316        0.969918                0.000000   

        mths_since_last_record  open_acc   pub_rec  revol_bal  revol_util  \
203160           -1.283214e-15 -0.806817 -0.330068  -0.349418    0.359436   
203161           -1.283214e-15  0.411787 -0.330068  -0.472140   -0.484567   
203162            7.843371e-01  0.005586  1.916061  -0.127374    0.315015   
203163           -1.283214e-15 -1.009918 -0.330068  -0.354523    0.153483   
203164            1.326126e+00  0.614888  1.916061  -0.057572    0.468470   

        total_acc  collections_12_mths_ex_med  mths_since_last_major_derog  \
203160  -1.444565                   -0.078593                    -0.007868   
203161   1.356033                   -0.078593                    -4.118160   
203162  -0.765633                   -0.078593                    -0.007868   
203163  -1.020232                   -0.078593                    -0.007868   
203164   0.083033                   -0.078593                    -0.007868   

        acc_now_delinq  tot_coll_amt  tot_cur_bal  open_acc_6m  open_il_6m  \
203160       -0.053128     -0.008119    -0.328050    -0.016454   -0.016307   
203161       -0.053128     -0.008119     0.452955    -0.016454   -0.016307   
203162       -0.053128     -0.003207    -0.820241    -0.016454   -0.016307   
203163       -0.053128     -0.008119     0.538710    -0.016454   -0.016307   
203164       -0.053128     -0.008119    -0.816277    -0.016454   -0.016307   

        open_il_12m  open_il_24m  mths_since_rcnt_il  total_bal_il   il_util  \
203160    -0.013663    -0.016004           -0.013738     -0.015985 -0.020929   
203161    -0.013663    -0.016004           -0.013738     -0.015985 -0.020929   
203162    -0.013663    -0.016004           -0.013738     -0.015985 -0.020929   
203163    -0.013663    -0.016004           -0.013738     -0.015985 -0.020929   
203164    -0.013663    -0.016004           -0.013738     -0.015985 -0.020929   

        open_rv_12m  open_rv_24m  max_bal_bc  all_util  total_rev_hi_lim  \
203160    -0.016504    -0.017879    -0.01846 -0.022036          0.000000   
203161    -0.016504    -0.017879    -0.01846 -0.022036         -0.603116   
203162    -0.016504    -0.017879    -0.01846 -0.022036         -0.350905   
203163    -0.016504    -0.017879    -0.01846 -0.022036         -0.595355   
203164    -0.016504    -0.017879    -0.01846 -0.022036         -0.315984   

          inq_fi  total_cu_tl  inq_last_12m  term_ 36 months  term_ 60 months  \
203160 -0.013258    -0.010854     -0.014543         0.549118        -0.549118   
203161 -0.013258    -0.010854     -0.014543         0.549118        -0.549118   
203162 -0.013258    -0.010854     -0.014543         0.549118        -0.549118   
203163 -0.013258    -0.010854     -0.014543         0.549118        -0.549118   
203164 

# 2. Split Dataset

##### Non-Scaled

In [83]:
X_train, X_val, y_train, y_val = train_test_split(X, y)
print ("X_train Shape :{} , y_train Shape:{}". format(X_train.shape, y_train.shape))
print ("X_val Shape :{} , y_val Shape:{}". format(X_val.shape, y_val.shape))

X_train Shape :(152373, 141) , y_train Shape:(152373,)
X_val Shape :(50792, 141) , y_val Shape:(50792,)


##### Scaled

In [84]:
X_train_s, X_val_s, y_train_s, y_val_s = train_test_split(X_scaled, y)
print ("X_train Shape :{} , y_train Shape:{}". format(X_train_s.shape, y_train_s.shape))
print ("X_val Shape :{} , y_val Shape:{}". format(X_val_s.shape, y_val_s.shape))

X_train Shape :(152373, 141) , y_train Shape:(152373,)
X_val Shape :(50792, 141) , y_val Shape:(50792,)


# 3. Model Selection

### (2) Dataframe sampling and shuffle

In [16]:
def shuffle_data(X_0, X_1): # select same number of sample with X_1 from X_0, and shuffle
    
    X_sample = X_0.sample(n=45936)
    
    X = pd.concat([X_sample, X_1])
    
    X = X.sample(frac=1).reset_index(drop=True)
    
    y = X['loan_status']
    X = X.drop('loan_status', axis=1)
    
    return X, y

### (3) Cross Validation

In [67]:
def cross(model, X_0, X_1, thres):
    
    X, y = shuffle_data(X_0, X_1)
    
    train_list = []
    test_list = []
    f1_list =[]
    auc_list=[]
    
     # 5 times modeling and take score
    for i in range(5):
        X_train, X_val, y_train, y_val = train_test_split(X, y)
        
        # Training
        fitted_model = model.fit(X_train, y_train)
        
        # accuracy
        train_list.append(model.score(X_train, y_train))
        test_list.append(model.score(X_val, y_val))
        
        # AUC $ ROC
        auc_list.append(roc_auc_score(y_val, model.predict(X_val)))
        #f1
        y_pred = model.predict_proba(X_val)
        y_true = y_val
        pred_random=[]

        THRESHOLD = thres
        for val in y_pred:
            if val[1] > THRESHOLD:
                pred_random.append(1)
            elif val[1] <= THRESHOLD:
                pred_random.append(0)
        f1_list.append(f1_score(y_true, pred_random))

   
    print('train score : {} & mean : {}'.format(train_list, np.array(train_list).mean()))
    print('test score: {}  & mean : {} '.format(test_list, np.array(test_list).mean()))
    print('f1 score : {} & mean : {}'.format(f1_list, np.array(f1_list).mean()))
    print('auc&roc score : {} & mean : {}'.format(auc_list, np.array(auc_list).mean()))
    
    
    return fitted_model, X_train 

### Test Score

In [41]:
def test (model, X, y, thres):
    
    # accuracy 
    test_accuracy = model.score(X, y)
     
    #f1
    y_pred = model.predict_proba(X)
    y_true = y
    pred_random=[]

    THRESHOLD = thres
    for val in y_pred:
        if val[1] > THRESHOLD:
            pred_random.append(1)
        elif val[1] <= THRESHOLD:
            pred_random.append(0)
    test_f1 = f1_score(y_true, pred_random)

     
    
    print('test score: {}'.format(test_accuracy))
    print('auc&roc score : {} '.format(roc_auc_score(y, model.predict(X))))
    print('f1 score : {} '.format(test_f1))
  

## 3-1) Random Forest

### *Predicting overfiting, tuning hypter parameter

In [21]:
model_random2 = RandomForestClassifier( random_state=2, n_estimators=500, n_jobs=-1, max_features='sqrt',oob_score = 'TRUE')


In [26]:
model_random_final= cross(model_random2, X_0, X_1, 0.35)

train score : [1.0, 1.0, 1.0, 1.0, 1.0] & mean : 1.0
test score: [0.66457680250783702, 0.66004876349703934, 0.65991814698711249, 0.66283524904214564, 0.66087600139324276]  & mean : 0.6616509926854756 
f1 score : [0.70339495798319329, 0.70520542413943932, 0.69770598615447266, 0.69694819020581966, 0.69526005289211357] & mean : 0.6997029222750076
auc&roc score : [0.66452068899785643, 0.66002020672339379, 0.65994015508939563, 0.66288743200649969, 0.66090039643121301] & mean : 0.6616537758496717


In [81]:
col_name = X.drop('loan_status', axis=1)

importances = model_random_final.feature_importances_

impo_feature = pd.DataFrame(importances, index=col_name.columns).sort_values(0, ascending=False)
impo_feature.head()

0
int_rate    0.055708
dti         0.052335
annual_inc  0.046392
revol_util  0.045317
sub_grade   0.044754

#### Looks very good hyperparameter


## 3-2) GradientBoosting

* Gradient Boosting takes too long time, gridsearch will not used. Instead hyper parameter will be tuning by hand

param = [{'n_estimators':[50, 100, 300, 500],
              'learning_rate':[0.1, 0.05, 0.01],
              'max_depth':[5, 10, 30]
             }]

#### Non-scaled

In [34]:

model_gradient= GradientBoostingClassifier(n_estimators=200, learning_rate=0.05, max_depth=5 )



In [35]:
model_gradient_final= cross(model_gradient, X_0, X_1, 0.24)

train score : [0.68607047486357831, 0.68730407523510972, 0.68953906884941363, 0.6880006966213863, 0.68742017879948913] & mean : 0.6876668988737954
test score: [0.66044061302681989, 0.65726227795193315, 0.65787182166492508, 0.6571316614420063, 0.66074538488331591]  & mean : 0.6586903517938001 
f1 score : [0.69061913118438456, 0.68888682573577187, 0.69345558598196988, 0.6890928256960378, 0.68895724392494617] & mean : 0.6902023225046221
auc&roc score : [0.66036652730344669, 0.65725689592076664, 0.65777320215115798, 0.65713430719867016, 0.66071341438137443] & mean : 0.6586488693910831


In [80]:
col_name = X.drop('loan_status', axis=1)

importances = model_gradient_final.feature_importances_

impo_feature = pd.DataFrame(importances, index=col_name.columns).sort_values(0, ascending=False)
impo_feature.head()

0
sub_grade         0.109854
annual_inc        0.069853
tot_cur_bal       0.059964
dti               0.057134
total_rev_hi_lim  0.052703

#### Scaled

In [ ]:

model_gradient= GradientBoostingClassifier(n_estimators=100, learning_rate=0.05, max_depth=5 )
model_gradient.fit(X_train_s, y_train_s)

In [ ]:
cross(model_gradient, X_scaled, y, 0.25)

## 3-3) Neural network models

random_states(0),alpha(0.0001 -> 1) :weight to be stronger,hidden_layer_sizes(100)

In [39]:
model_mlp = MLPClassifier(alpha=0.0001,max_iter=500, hidden_layer_sizes=200)
#model_mlp.fit(X_train, y_train)

In [42]:
model_nn_final= cross(model_mlp, X_0, X_1, 0.33)

train score : [0.55557006850110302, 0.53255253686288173, 0.51696563334494372, 0.59880413328689186, 0.58297050969464759] & mean : 0.5573725763380937
test score: [0.55725357018460464, 0.53073841866945315, 0.52237896203413448, 0.59543712991988851, 0.57911006617903171]  & mean : 0.5569836293974225 
f1 score : [0.67131557707828893, 0.67309149316604533, 0.67237061769616013, 0.64606720194832368, 0.52599568373553074] & mean : 0.6377681147248697
auc&roc score : [0.55447875555568893, 0.52964145099595716, 0.51861466107552945, 0.59568263886125483, 0.58005434362742014] & mean : 0.55569437002317


## 3-4) XG Boost

gamma, max_depth, lambda, alpha, normalize type

In [68]:

xgbr = xgb.XGBClassifier (n_estimators=30, gamma=3, max_depth=10, objective='binary:logistic', normalize_type='forest' )


In [73]:

X_0.sort_index(axis=1, inplace = True)
X_1.sort_index(axis=1, inplace = True)
X_test.sort_index(axis=1,inplace=True)

In [75]:
model_xg_final, X_train_xg= cross(xgbr, X_0, X_1, 0.33)

train score : [0.76471612678509227, 0.76298908626494832, 0.76535469638917919, 0.76541274817136884, 0.76346801346801352] & mean : 0.7643881342157204
test score: [0.65269070010449326, 0.64842389411354928, 0.65238592824799724, 0.65060083594566354, 0.65647857889237204]  & mean : 0.6521159874608151 
f1 score : [0.70075858397657698, 0.6962152059275083, 0.69466697951932421, 0.69602994252105332, 0.69552988448016073] & mean : 0.6966401192849248
auc&roc score : [0.65269277339022636, 0.64842285980982484, 0.65238381507210019, 0.65059600983961674, 0.65646985065609043] & mean : 0.6521130617535718


In [83]:
col_name = X.drop('loan_status', axis=1)

importances = model_xg_final.feature_importances_

impo_feature = pd.DataFrame(importances, index=col_name.columns).sort_values(0, ascending=False)
impo_feature.head()

0
veri_Verified  0.077888
home_OWN       0.069681
term           0.067708
pur_vacation   0.067314
issue_Jan      0.064157

## 3-5) Light gbm

In [45]:
gbm = lgb.LGBMClassifier(objective='regression',
                        num_leaves=200,
                        learning_rate=0.001,
                        n_estimators=3000,
                        # early_stopping_rounds=5
                       )



In [47]:
model_gbm_final = cross(gbm, X_0, X_1, 0.36)

train score : [0.74336758388482527, 0.74452861952861948, 0.74428189945431322, 0.74268547544409613, 0.7440496923255544] & mean : 0.7437826541274818
test score: [0.65952629745733193, 0.65782828282828287, 0.65325670498084287, 0.66000522466039713, 0.6566527342389411]  & mean : 0.6574538488331592 
f1 score : [0.69997243660418951, 0.69800559952991603, 0.69418737652074458, 0.70376240352811459, 0.69820317899101603] & mean : 0.6988261990347963
auc&roc score : [0.65956065090604821, 0.65783668500665382, 0.65334899155176873, 0.65995545239605025, 0.65661908792063106] & mean : 0.6574641735562304


In [84]:
col_name = X.drop('loan_status', axis=1)

importances = model_gbm_final.feature_importances_

impo_feature = pd.DataFrame(importances, index=col_name.columns).sort_values(0, ascending=False)
impo_feature.head()

0
annual_inc     0.073342
dti            0.072496
revol_util     0.066707
credit_period  0.058479
tot_cur_bal    0.057720

# 4) Applying to Testset

### 1) Random Forest

In [38]:
test(model_random_final, X_test, y_test, 0.35)

test score: 0.6456690587992086
auc&roc score : 0.656774332377025 
f1 score : 0.4235437819680892 


### 2) Gradient Boosting

In [48]:
test(model_gradient_final, X_test, y_test, 0.25)

test score: 0.6570095389976669
auc&roc score : 0.6582475146334387 
f1 score : 0.3997114978974729 


### 3) Neural network model

In [49]:
test(model_nn_final, X_test, y_test, 0.33)

test score: 0.6462597088095449
auc&roc score : 0.5806297067130928 
f1 score : 0.37103221255584296 


### 4) XGBoost

In [78]:
test(model_xg_final, X_test, y_test, 0.33)

test score: 0.654957030211748
auc&roc score : 0.6550374639133867 
f1 score : 0.42015710776136606 


### 5) LightGBM

In [52]:
test(model_gbm_final, X_test, y_test, 0.36)

test score: 0.6529045214258291
auc&roc score : 0.6562043019173878 
f1 score : 0.4350461047212538 


# 5) Conclusion

1.  Selected Model
> 1.  Light GBM
<br> - More faster than XGBoost and better performance
<br> - Having higher score than Random Forest
<br>
> 2. Infortance Features
<br> - annual_inc :	0.0733
<br> - dti : 0.0725
<br> - revol_util : 0.0667
<br> - credit_period : 0.0585
<br> - tot_cur_bal : 0.0577
<br><br>
2.  Necessity of more Non-identification infomation.
> 
<br> According to Finance Report by Dr. Asim Khwaja from Havard University, not only finance information, but also non-finance information help predict loan payment as well, like handwritin on application, sns, insurance.
<br> I'm interested in analys with this data later, it looks very helpul but hard to get